[Issue#3976] [component/client] Adding config for ReaderImpl

ConcurrencyPractitioner opened the pull request:
Fixes #3976

### Motivation

We wish to add a method which  would allow ReaderBuilderImpl to create a reader which starts reading from a specified timestamp by the user. 

### Modifications

Added a configuration and method to allow for a Reader created by ReaderBuilderImpl to start processing from a selected timestamp.

lovelle commented on pulsar-client/src/main/java/org/apache/pulsar/client/impl/ReaderImpl.java:
```diff
@@ -87,7 +91,13 @@ public void reachedEndOfTopic(Consumer<T> consumer) {
                 client.getConfiguration().getDefaultBackoffIntervalNanos(), client.getConfiguration().getMaxBackoffIntervalNanos());
-        
+        try {
+            if (readerConfiguration.getTimestamp().longValue() != -1) {
```
Explicit unwrapping is no longer necessary since Java 5

ConcurrencyPractitioner commented:
@sijie  I noticed that usually the user would either use startMessageId(long timestamp) or startMessageId(MessageId id) but not both. So should we make it so that we need to call only one of the methods to set the message id at which the user would need to start? Currently, it is required that MessageId id be a non-null parameter, so if the user wants to use startMessageId(long timestamp), they would also have to call startMessageId(MessageId id) with some dummy variable (id) as well.

ConcurrencyPractitioner commented:
The test added is broken, although I couldn't pinpoint the exact cause. The location where the consumer seeks to the timestamp the user selects might be incorrect, or is ineffective, seeing that when the reader polls messages, we still get my-message-0 as the first message instead of my-message-3. I will look into it.

sijie commented on pulsar-broker/src/test/java/org/apache/pulsar/client/api/TopicReaderTest.java:
```diff
@@ -86,6 +87,33 @@ public void testSimpleReader() throws Exception {
+         for (int i = 0; i < 10; i++) {
+            String message = "my-message-" + i;
+            producer.send(message.getBytes());
+            Thread.sleep(1000);
```
why do you add `sleep` here?

ConcurrencyPractitioner replied:
My idea was to use the sleep so that we could space the publish times of the messages 1000ms apart (although that might or might have not worked).

sijie commented on pulsar-broker/src/test/java/org/apache/pulsar/client/api/TopicReaderTest.java:
```diff
@@ -86,6 +87,33 @@ public void testSimpleReader() throws Exception {
+                .create();
+         for (int i = 0; i < 10; i++) {
+            String message = "my-message-" + i;
+            producer.send(message.getBytes());
```
use `UTF_8`

sijie commented on pulsar-broker/src/test/java/org/apache/pulsar/client/api/TopicReaderTest.java:
```diff
@@ -86,6 +87,33 @@ public void testSimpleReader() throws Exception {
+        Message<byte[]> msg = null;
+        Set<String> messageSet = Sets.newHashSet();
+        for (int i = 3; i < 10; i++) {
+            msg = reader.readNext(1, TimeUnit.SECONDS);
```
why do you have a timeout here? If the CI environment is slow, readNext will fail, no?

sijie commented on pulsar-broker/src/test/java/org/apache/pulsar/client/api/TopicReaderTest.java:
```diff
@@ -86,6 +87,33 @@ public void testSimpleReader() throws Exception {
+        for (int i = 3; i < 10; i++) {
+            msg = reader.readNext(1, TimeUnit.SECONDS);
+
+            String receivedMessage = new String(msg.getData());
```
use UTF_8 encoding

sijie commented on pulsar-client-api/src/main/java/org/apache/pulsar/client/api/ReaderBuilder.java:
```diff
@@ -121,6 +121,13 @@
+     * @param timestamp the timestamp where the reader will be initially positioned on
+     * @return the reader builder instance
+     */
+    ReaderBuilder<T> startMessageId(long timestamp);
```
I would suggest calling it `startPublishTime` because `timestamp` is not a message id.

sijie commented on pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java:
```diff
@@ -220,13 +220,13 @@ protected ConsumerImpl(PulsarClientImpl client, String topic, ConsumerConfigurat
-        	               .useUserConfiguredIntervals(backoffIntervalNanos, 
-        	                                           maxBackoffIntervalNanos)
-        	               .create(),
+                      new BackoffBuilder()
```
try to avoid formatting the code as reformatting will make reviewers harder to review your code.

sijie commented on pulsar-client/src/main/java/org/apache/pulsar/client/impl/ReaderBuilderImpl.java:
```diff
@@ -87,8 +90,13 @@ private ReaderBuilderImpl(PulsarClientImpl client, ReaderConfigurationData<T> co
-            return FutureUtil
-                    .failedFuture(new IllegalArgumentException("Start message id must be set on the reader builder"));
+            if (conf.getTimestamp() != -1) {
+                log.warn("Setting messageId value since timestamp value is {}", conf.getTimestamp());
```
why the log level here is `warn`?

sijie commented on pulsar-client/src/main/java/org/apache/pulsar/client/impl/ReaderBuilderImpl.java:
```diff
@@ -87,8 +90,13 @@ private ReaderBuilderImpl(PulsarClientImpl client, ReaderConfigurationData<T> co
         if (conf.getStartMessageId() == null) {
-            return FutureUtil
-                    .failedFuture(new IllegalArgumentException("Start message id must be set on the reader builder"));
+            if (conf.getTimestamp() != -1) {
```
I think you should `Long` to present whether this field is set or not.

sijie commented on pulsar-client/src/main/java/org/apache/pulsar/client/impl/conf/ReaderConfigurationData.java:
```diff
@@ -32,6 +32,7 @@
 
     private String topicName;
     private MessageId startMessageId;
+    private Long timestamp = -1L;
```
don't use `-1L` for indicating whether this setting is set or not.

sijie replied:
call it `publishTime` or `publishTimestamp`. because `timestamp` is an ambiguous term, it has multiple meanings in pulsar - publish time or event time.

sijie commented on pulsar-client/src/main/java/org/apache/pulsar/client/impl/ReaderImpl.java:
```diff
@@ -87,7 +91,13 @@ public void reachedEndOfTopic(Consumer<T> consumer) {
                 client.getConfiguration().getDefaultBackoffIntervalNanos(), client.getConfiguration().getMaxBackoffIntervalNanos());
-        
+        try {
+            if (readerConfiguration.getTimestamp() != -1) {
```
use `Null` for checking if the value is set or not.

ConcurrencyPractitioner commented:
@sijie I want to confirm that seek() should be used for this configuration right?

sijie commented:
@ConcurrencyPractitioner yes.

ConcurrencyPractitioner commented:
So while investigating, a couple of things was revealed. There is actually two approaches to getting to the right initial startPublishTimestamp:
1. Add a new parameter or field  so the initial subscription command (i.e. ```Commands.newSubscribe```) would entail broker to seek to startPublishTimestamp directly.
2. We call seek() from the client side instead to get to startPublishTimestamp.

@sijie  which one do you think is preferable?

lovelle commented:
Hi! In my humble opinion I think I would vote for option number 2 with proper documentation to avoid misuse from the user. Mainly for simplicity reasons.

Also, I'm testing 2nd option and facing a non deterministic issue about unlucky timing being disconnected and not connected after seek() operation.
I'm not sure if this is from the fact of having a non-durable consumer or what, I will keep looking around this issue.

sijie commented:
I would also prefer option 2 as well for simplicity reasons.

ConcurrencyPractitioner commented:
Alright, so I ran the test on local and here was a segment of the log for the test I added:

```
Seeking to time: 1554862866386
19:21:13.441 [TestNG-method=testReaderAtSelectedTimestamp-1:org.apache.pulsar.client.impl.ReaderBuilderImpl@94] INFO  org.apache.pulsar.client.impl.ReaderBuilderImpl - Setting messageId value since timestamp value is 1554862866386
19:21:13.450 [pulsar-web-149-6:org.eclipse.jetty.server.Slf4jRequestLog@60] INFO  org.eclipse.jetty.server.RequestLog - 127.0.0.1 - - [09/Apr/2019:19:21:13 -0700] "GET /admin/v2/persistent/my-property/my-ns/testReaderAtSelectedTimestamp/partitions HTTP/1.1" 200 16 "-" "Pulsar-Java-v2.4.0-SNAPSHOT" 6
19:21:13.458 [pulsar-web-149-7:org.eclipse.jetty.server.Slf4jRequestLog@60] INFO  org.eclipse.jetty.server.RequestLog - 127.0.0.1 - - [09/Apr/2019:19:21:13 -0700] "GET /lookup/v2/topic/persistent/my-property/my-ns/testReaderAtSelectedTimestamp HTTP/1.1" 200 132 "-" "Pulsar-Java-v2.4.0-SNAPSHOT" 6
19:21:13.459 [pulsar-client-io-155-1:org.apache.pulsar.client.impl.ConsumerImpl@493] INFO  org.apache.pulsar.client.impl.ConsumerImpl - [persistent://my-property/my-ns/testReaderAtSelectedTimestamp][reader-e722c2aff4] Subscribing to topic on cnx [id: 0xf5d3a09a, L:/127.0.0.1:59183 - R:localhost/127.0.0.1:15075]
19:21:13.461 [pulsar-io-142-1:org.apache.pulsar.broker.service.ServerCnx@641] INFO  org.apache.pulsar.broker.service.ServerCnx - [/127.0.0.1:59183] Subscribing on topic persistent://my-property/my-ns/testReaderAtSelectedTimestamp / reader-e722c2aff4
19:21:13.461 [pulsar-io-142-1:org.apache.pulsar.broker.service.persistent.PersistentTopic@624] INFO  org.apache.pulsar.broker.service.persistent.PersistentTopic - [persistent://my-property/my-ns/testReaderAtSelectedTimestamp][reader-e722c2aff4] Creating non-durable subscription at msg id -1:-1:-1:-1
19:21:13.462 [pulsar-io-142-1:org.apache.bookkeeper.mledger.impl.NonDurableCursorImpl@54] INFO  org.apache.bookkeeper.mledger.impl.NonDurableCursorImpl - [my-property/my-ns/persistent/testReaderAtSelectedTimestamp] Created non-durable cursor read-position=3:0 mark-delete-position=3:-1
19:21:13.462 [pulsar-io-142-1:org.apache.pulsar.broker.service.persistent.PersistentTopic@648] INFO  org.apache.pulsar.broker.service.persistent.PersistentTopic - Cursor's name is: null
19:21:13.463 [pulsar-io-142-1:org.apache.bookkeeper.mledger.impl.ManagedCursorImpl@1789] INFO  org.apache.bookkeeper.mledger.impl.ManagedCursorImpl - [my-property/my-ns/persistent/testReaderAtSelectedTimestamp-null] Rewind from 3:0 to 3:0
19:21:13.463 [pulsar-io-142-1:org.apache.pulsar.broker.service.persistent.PersistentTopic@571] INFO  org.apache.pulsar.broker.service.persistent.PersistentTopic - [persistent://my-property/my-ns/testReaderAtSelectedTimestamp][reader-e722c2aff4] Created new subscription for 0
19:21:13.463 [pulsar-io-142-1:org.apache.pulsar.broker.service.ServerCnx@701] INFO  org.apache.pulsar.broker.service.ServerCnx - [/127.0.0.1:59183] Created subscription on topic persistent://my-property/my-ns/testReaderAtSelectedTimestamp / reader-e722c2aff4
19:21:13.464 [pulsar-client-io-155-1:org.apache.pulsar.client.impl.ConsumerImpl@588] INFO  org.apache.pulsar.client.impl.ConsumerImpl - [persistent://my-property/my-ns/testReaderAtSelectedTimestamp][reader-e722c2aff4] Subscribed to topic on localhost/127.0.0.1:15075 -- consumer: 0
19:21:13.468 [TestNG-method=testReaderAtSelectedTimestamp-1:org.apache.pulsar.client.impl.ConsumerImpl@1376] INFO  org.apache.pulsar.client.impl.ConsumerImpl - [persistent://my-property/my-ns/testReaderAtSelectedTimestamp][reader-e722c2aff4] Seek subscription to publish time 1554862866386
19:21:13.473 [pulsar-io-142-1:org.apache.pulsar.broker.service.ServerCnx@210] WARN  org.apache.pulsar.broker.service.ServerCnx - [/127.0.0.1:59183] Got exception NullPointerException : null
java.lang.NullPointerException: null
        at java.net.URLDecoder.decode(URLDecoder.java:136) ~[?:1.8.0_191]
        at org.apache.pulsar.common.util.Codec.decode(Codec.java:45) ~[pulsar-common-2.4.0-SNAPSHOT.jar:2.4.0-SNAPSHOT]
        at org.apache.pulsar.broker.service.persistent.PersistentMessageFinder.<init>(PersistentMessageFinder.java:55) ~[classes/:?]
        at org.apache.pulsar.broker.service.persistent.PersistentSubscription.resetCursor(PersistentSubscription.java:335) ~[classes/:?]
        at org.apache.pulsar.broker.service.ServerCnx.handleSeek(ServerCnx.java:1153) ~[classes/:?]
        at org.apache.pulsar.common.api.PulsarDecoder.channelRead(PulsarDecoder.java:228) ~[pulsar-common-2.4.0-SNAPSHOT.jar:2.4.0-SNAPSHOT]
```

What I found curious in the log output was that a NonDurableCursor was created, but its name was null. But later, a decoder used by PersistentMessageFinder attempts to decode the name fails because the name has a null vale So its been bugging me since this shouldn't happen. @sijie Do you have any pointers on why this might be the case?

ConcurrencyPractitioner commented:
One more thing. I have tried to turn on DEBUG mode for log4j when running the tests. What I did was I copied the log4j.properties of pulsar-proxy to pulsar-broker/src/main/resources/log4j.properties and changed the first line of the file to DEBUG, CONSOLE. Unfortunately, that did not work. Is there a way where we could turn on DEBUG mode while running TopicReaderTest?

ConcurrencyPractitioner commented:
Alright, I'm not sure if option 2 still works. From log after test rerun:

```
21:36:55.342 [bookkeeper-ml-workers-OrderedExecutor-1-0:org.apache.pulsar.broker.service.persistent.PersistentMessageFinder@99] INFO  org.apache.pulsar.broker.service.persistent.PersistentMessageFinder - [persistent://my-property/my-ns/testReaderAtSelectedTimestamp][reader-3f11d6ab7e] Found position 3:2 closest to provided timestamp 1554871008242
21:36:55.342 [bookkeeper-ml-workers-OrderedExecutor-1-0:org.apache.pulsar.broker.service.Consumer@369] INFO  org.apache.pulsar.broker.service.Consumer - Disconnecting consumer: Consumer{subscription=PersistentSubscription{topic=persistent://my-property/my-ns/testReaderAtSelectedTimestamp, name=reader-3f11d6ab7e}, consumerId=0, consumerName=2a5c6, address=/127.0.0.1:61416}
21:36:55.344 [pulsar-client-io-187-1:org.apache.pulsar.client.impl.ClientCnx@600] INFO  org.apache.pulsar.client.impl.ClientCnx - [localhost/127.0.0.1:15087] Broker notification of Closed consumer: 0
21:36:55.344 [pulsar-client-io-187-1:org.apache.pulsar.client.impl.ConnectionHandler@109] INFO  org.apache.pulsar.client.impl.ConnectionHandler - [persistent://my-property/my-ns/testReaderAtSelectedTimestamp] [reader-3f11d6ab7e] Closed connection [id: 0x2ac3250a, L:/127.0.0.1:61416 - R:localhost/127.0.0.1:15087] -- Will try again in 0.1 s
21:36:55.345 [bookkeeper-ml-workers-OrderedExecutor-1-0:org.apache.pulsar.broker.service.AbstractDispatcherSingleActiveConsumer@159] INFO  org.apache.pulsar.broker.service.AbstractDispatcherSingleActiveConsumer - Removing consumer Consumer{subscription=PersistentSubscription{topic=persistent://my-property/my-ns/testReaderAtSelectedTimestamp, name=reader-3f11d6ab7e}, consumerId=0, consumerName=2a5c6, address=/127.0.0.1:61416}
21:36:55.345 [bookkeeper-ml-workers-OrderedExecutor-1-0:org.apache.pulsar.broker.service.persistent.PersistentSubscription@499] INFO  org.apache.pulsar.broker.service.persistent.PersistentSubscription - [persistent://my-property/my-ns/testReaderAtSelectedTimestamp][reader-3f11d6ab7e] Successfully closed subscription [NonDurableCursorImpl{ledger=my-property/my-ns/persistent/testReaderAtSelectedTimestamp, ackPos=3:-1, readPos=3:10}]
21:36:55.346 [bookkeeper-ml-workers-OrderedExecutor-1-0:org.apache.pulsar.broker.service.persistent.PersistentSubscription@406] INFO  org.apache.pulsar.broker.service.persistent.PersistentSubscription - [persistent://my-property/my-ns/testReaderAtSelectedTimestamp][reader-3f11d6ab7e] Successfully disconnected consumers from subscription, proceeding with cursor reset
21:36:55.346 [bookkeeper-ml-workers-OrderedExecutor-1-0:org.apache.bookkeeper.mledger.impl.ManagedCursorImpl@822] INFO  org.apache.bookkeeper.mledger.impl.ManagedCursorImpl - [my-property/my-ns/persistent/testReaderAtSelectedTimestamp] Initiate reset position to 3:2 on cursor reader-3f11d6ab7e
21:36:55.346 [bookkeeper-ml-workers-OrderedExecutor-1-0:org.apache.bookkeeper.mledger.impl.ManagedCursorImpl$7@861] INFO  org.apache.bookkeeper.mledger.impl.ManagedCursorImpl - [my-property/my-ns/persistent/testReaderAtSelectedTimestamp] reset position to 3:2 before current read position 3:10 on cursor reader-3f11d6ab7e
21:36:55.346 [bookkeeper-ml-workers-OrderedExecutor-1-0:org.apache.pulsar.broker.service.ServerCnx@1154] INFO  org.apache.pulsar.broker.service.ServerCnx - [/127.0.0.1:61416] [persistent://my-property/my-ns/testReaderAtSelectedTimestamp][reader-3f11d6ab7e] Reset subscription to publish time 1554871008242
21:36:55.347 [pulsar-client-io-187-1:org.apache.pulsar.client.impl.ConsumerImpl@1379] INFO  org.apache.pulsar.client.impl.ConsumerImpl - [persistent://my-property/my-ns/testReaderAtSelectedTimestamp][reader-3f11d6ab7e] Successfully reset subscription to publish time 1554871008242
!!!!!!!!! FAILURE-- [TestClass name=class org.apache.pulsar.client.api.TopicReaderTest].null([])-------
```

So the seek was supposed to have worked judging from the log ouput. And the cursor was reset from 3:0 to 3:2. Although obviously, the test still failed afterwards.

lovelle commented:
@sijie @ConcurrencyPractitioner please take a look at pull #4031 , thanks!

ConcurrencyPractitioner commented:
Alright, so I've been working on approach 1 since approach 2 didn't seem to be working. So I have added a new field to PulsarApi.proto and I wanted to generate PulsarApi.java accordingly. However, that was where I ran into some problems since I was unable to perform the second step. I have added a ```protobuf``` dependency and ```protobuf:compile``` plugin goal to pulsar-common/pom.xml but no changes were made to PulsarApi.java. So right now, I probably need some advice on how to generate the PulsarApi.java file.

lovelle commented:
In order to generate `PulsarApi.java` try with `generate_protobuf_docker.sh` script, it always worked well for me.

I'm still think that path 1 is not quite a good idea.

I wouldn't say that pull #4031 isn't working, the only case we didn't cover so far is a reset() by time in the middle of a queue, in fact, the `reset()` is happening but not at the expected position, is happening on a previous position, which would result in a duplicates messages which I don't think is a critical situation, and also I think that normal Consumers are facing this issue too.

@sijie Any thought on this?

ConcurrencyPractitioner commented:
Oh, you might've a point. (BTW, just wanted to see if the ```generate_protobuf_docker.sh``` works and it did. So thanks for the help on that. :))

ConcurrencyPractitioner commented:
Alright, so here is my current approach on the issue.

```resetCursor``` is known to cause the consumer to disconnect. That might've been the cause of why we are starting to process from my-message-0 despite the seek call. So what I have done is that during the ```subscribe``` command, I match the start publish time with the message id ```startId``` that is the closest to it, and then I will call ```getNonDurableSubscription(MessageId startMessageId)``` to create the subscription where startMessageId equals ```startId```.

That is the basic approach that I'm currently doing.

ConcurrencyPractitioner commented:
@sijie Mind taking a look?

ConcurrencyPractitioner commented:
@sijie Mind taking a look?

ConcurrencyPractitioner commented:
Actually, @sijie Do you mind if you take a look at lovelle's PR first? Once his is in, I will build on top of his approach and finish up this feature. (My current approach is flawed, so there's no need to look at my PR at the moment).

sijie commented:
@ConcurrencyPractitioner sure. doing it.

ConcurrencyPractitioner commented:
Retest this please.
